### PR TITLE
[REG-1666] Gameplay session docs

### PR DESCRIPTION
**Pull request by vontell**
Loom video with an overview: https://www.loom.com/share/b6e0ceeaa2e84d3482d65768f6f3e984?sid=ebba2aa5-513a-48db-bbab-6e25957c9e58

Note to reviewers: If you find yourself making a ton of grammatical corrections, I'd prefer if you can just pull down the branch and push those changes directly... I trust your judgement!

This PR introduces documentation for the gameplay session concept. Note that this does not yet cover the web UI side of things yet, and only covers the SDK.

I structured this into three pages:
- An overview
- Getting started and basic usage
- A technical reference (not only for power users, but for ourselves)

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    }
+  },
+  "position": {
+    "x": number, // A float representing the x coordinate of the position of the object
+    "y": number, // A float representing the y coordinate of the position of the object
+    "z": number // A float representing the z coordinate of the position of the object
+  },
+  "rotation": {
+    "x": number, // A float representing the x component of the rotation of the object
+    "y": number, // A float representing the y component of the rotation of the
```
Comment (vontell) [inline docs/gameplay-sessions/gameplay-sessions-reference.mdx:114]: Noticed there may be some issues in these comments. Maybe I should just remove a lot of these comments... some of these fields are pretty clear

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+#### KeyboardInput Schema
+
+Within the `inputs` field, an array of objects represents the keyboard input on this tick. Each object has the following
+schema:
+
+```json
+{
+  "startTime": number, // A float representing the time in seconds since the start of the session when this input started
+  "key": string // A string representing the key that was pressed
+}
```
Comment (vontell) [inline docs/gameplay-sessions/gameplay-sessions-reference.mdx:133]: As noted in the video, this is not correct (was waiting for another PR, but will fix this soon)

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx (hunk 1 above)
Comment (addisonbgross) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx:114]: I think removing these comments is a good idea 👍🏻 
This part of the docs is going to change constantly I think

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+By default, all data is stored to a `unity_videos` directory within your home directory. This can be changed within the
+**State Recordings Directory** field.
+
+TO DISCUSS IN PR COMMENTS: How does this work when multiple people are using the project on different machines?
+
+![Change the location where data is saved](img/base_options_location.png)
+
+### Input threshold options
+
+In some cases, different devices may report spammy key clicks when a key is actually being held down. This threshold
```
[the remark this answers is not in the document]
Comment (addisonbgross) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx:68]: I like `key presses`

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+within the SDK itself, you may be interested in viewing and understanding the data directly. This document provides a
+technical overview of the Gameplay Sessions system, as well as more technical details and FAQ.
+
+## Overview of the Schema
+
+### Folder and File Structure
+
+When a Gameplay Session is recorded, all data is stored on disk within two folders:
+
+* `~/unity_videos/run_{#}/screenshots` - A folder with screenshots for each recorded tick of the session. Each
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Ah good catch, updated.

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+* `~/unity_videos/run_{#}/data` - A folder with the state and input data for each recorded tick of the session. Each
+  tick is stored in a JSON file named with the format `{tick#}.json`.
+
+After a recording is finished, you will not see these folders, because they will be compressed into zip files
+(`data.zip` and `screenshots.zip` within the `run_{#}` folder).
+
+Additionally, there is a `thumbnail.jpg` file stored in this run folder, which is the screenshot from the middle of
+your gameplay session.
+
+### Structure of {tick}.json Files
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Added

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+Each `{tick}.json` file contains the following information:
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Added

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx (hunk 1 above)
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Removed!

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+  "backButton": boolean, // A boolean representing whether the back mouse button was pressed
+  "scroll": {
+    "x": number, // A float representing the x component of the scroll
+    "y": number // A float representing the y component of the scroll
+  },
+  "clickedObjectIds": [number] // An array of integers representing the unique identifiers of the objects clicked
+}
+```
+
+### Key Frames
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: With the new structure, let me know if this makes more sense. Happy to add a sentence or two if there is something specific you'd like to say!

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+
+Key frames are a way to mark important moments in your gameplay session. They are represented as an array of strings in
+the `keyFrame` field of the `{tick}.json` file. The possible key frames are:
+
+* `FIRST_FRAME` - Indicates that this tick is the first tick of the gameplay session.
+* `SCENE` - Indicates that this tick is the first tick of a scene that is different from the previous tick's scene.
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Fixed, although intellij grammar check told me to keep the `from` vs `than`

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+* `GAME_ELEMENT` - Indicates that a visible game element was changed this tick.
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Fixed

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+* `GAME_ELEMENT_RENDERER_COUNT` - Indicates that a visible game element changed its number of renderers this tick.
+* `UI_ELEMENT` - Indicates that a visible UI element was changed this tick.
+* `UI_PIXELHASH` - Indicates that the screen pixels changed this tick.
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Added

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+### Key Frames
+
+Key frames are a way to mark important moments in your gameplay session. They are represented as an array of strings in
+the `keyFrame` field of the `{tick}.json` file. The possible key frames are:
+
+* `FIRST_FRAME` - Indicates that this tick is the first tick of the gameplay session.
+* `SCENE` - Indicates that this tick is the first tick of a scene that is different from the previous tick's scene.
+* `GAME_ELEMENT` - Indicates that a visible game element was changed this tick.
+* `GAME_ELEMENT_RENDERER_COUNT` - Indicates that a visible game element changed its number of renderers this tick.
+* `UI_ELEMENT` - Indicates that a visible UI element was changed this tick.
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Added

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+## How GameObjects are scraped
+
+## How performance information is scraped
+
+Performance information, such as fps, render time, and object stats are either calculated frame-by-frame, or use
+the [UnityStats](https://stephenhodgson.github.io/UnityCsReference/api/UnityEditor.UnityStats.html) class. Since
+UnityStats is an editor-only feature, some performance information, specifically that of `engineStats`, is only
+available when running in the Unity Editor.
+
+## How input is scraped
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: Removed

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import OverlayPartial from '../partials/_add-overlay-partial.mdx';
+
+# Using Gameplay Sessions
+
+## Preparing the Overlay
+
+First, add the RGOverlayCanvas prefab to your scene, which is the UI that allows us to record and play gameplay
+sessions.
+
+<OverlayPartial />
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx:14]: 🎉

**Review by vontell — COMMENTED** [review]
Addressed comments! Back to you all @addisonbgross @abeizer @RG-nAmKcAz

File: docs/tutorials/building-your-first-bot.mdx [hunk 14]
```diff
@@ -69,33 +71,20 @@ resolve the new packages even after Unity recompiles the scripts.
...
-Regression Games logo will appear at the bottom of your Scene View. Click this logo to see a dropdown for your bots. You
-likely will not have any bots yet - that's alright, we will make one soon!
+Once the overlay has been added to your scene(s), you can click play to see it in action!
+Click the Regression Games logo at the bottom of the Scene View to open the overlay.
 
 :::info
 
-Your scene must have an EventSystem to make interaction with the overlay possible. You can add one by 
-right-clicking in your scene's Hierarchy and selecting **GameObject** > **UI** > **Event System**. 
+Your scene must have an EventSystem to make interaction with the overlay possible. You can add one by
```
Comment (RG-nAmKcAz) [inline docs/tutorials/building-your-first-bot.mdx:81]: a

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+TO DISCUSS IN PR COMMENTS: How does this work when multiple people are using the project on different machines?
+
+![Change the location where data is saved](img/base_options_location.png)
+
+### Input threshold options
+
+In some cases, different devices may report spammy key clicks when a key is actually being held down. This threshold
+defines a maximum amount of time that must exist between key clicks to be considered clicks rather than a key being
+held. This defaults to 0.1 seconds, and often should not have to be changed. If you do experience key click spamming,
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: This comment has Not been addressedThis comment has Not been addressed

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+In addition to ignore certain objects as described above, there are also a few other ways that this can be configured.
+
+#### Collapsing Renderers
+
+The **Collapse Renderers into Top Level Game Object** checkbox dictates how renderer game objects are collected as part
+of the game state. If true, all renderers are collapsed into their top level gameObject. If a gameObject hierarchy
+exists that has colliders/renderers/animators/etc at multiple levels, they will all be represented by a single entry in
+the state. This defaults to checked as it is normally desired to see each player, car, building, etc as a single entry
+in the state. However, it can be useful to set this to false in cases where you want to validate individual render
+bounds, colliders, rigid bodies, and more on individual armatures, weapons, or other components that are children in the
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: This comment has Not been addressed

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+hierarchy.
+
+#### Include Only On Camera Objects
+
+The **Include Only On Camera Objects** checkbox dictates whether the recorder should collect all objects in the
+scene as part of the game state, or only those that appear in the viewport of the camera. This defaults to checked, as
+in many cases, objects that are not on camera are not relevant to the test or playback. If you do need to see all
+objects in the scene, uncheck this field.
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: This comment has Not been addressed

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+run you'd like to use, select the `data.zip` file, and click "Load Replay".
+
+After the replay is selected, click the Play button underneath the Regression Games icon to begin the playback.
+
+![Open the Gameplay Session Selector](img/overlay_click_load.png)
+![Select the Gameplay Session](img/overlay_choose_session.png)
+
+## Advanced Options
+
+### Ignoring objects in your Scene
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: This comment has Not been addressed

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+1. Search for and add the `RG Ignore From State` component
+
+This will ignore all state from this game object as well as all children of this game object.
+
+### Choosing an FPS for recording
+
+The **Recording Min FPS** option tells the SDK the frequency at which it should record information from the game. For
+instance, if the FPS is set to `5`, then 5 times a second the SDK will scrape the game information.
+
+If the FPS is set `0`, then the SDK will only record information when a
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: This comment was only partially addressed, we still need a disclaimer that this value should normally stay as 0.

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]
Some of the original review comments still need addressing as noted

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+## Advanced Options
+
+### Ignoring objects in your Scene
+
+In certain scenarios, you may want to ignore certain objects in your recordings, such as regions of a level that are
+not relevant to the test or behavior of your recording. To ignore these game objects, follow these steps:
+
+1. Click "Add Component"
+1. Search for and add the `RG Ignore From State` component
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Fixed (I believe @abeizer's commit fixes this)

**Comment by vontell** [comment]
@RG-nAmKcAz apologies, for some reason those comments weren't showing up in the code view on GitHub (i think some got outdated from Abby's commit). Going through these now!

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx (hunk 20 above)
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Yes, I had already marked this point resolved

**Review by abeizer — APPROVED** [review]
Approving now since I made many changes to these docs yesterday

**Review by addisonbgross — APPROVED** [review]
Approved pending Zack's comments

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+scene as part of the game state, or only those that appear in the viewport of the camera. This defaults to checked, as
+in many cases, objects that are not on camera are not relevant to the test or playback. If you do need to see all
+objects in the scene, uncheck this field.
+
+#### Collect State from Behaviours
+
+The **Collect State from Behaviours** checkbox dictates whether the recorder should collect fields and property
+values from Behaviours attached to the game objects. For recording playback, this is not always needed, but for testing
+specific states within your game objects, this is extremely useful. This is unchecked by default, and note that
+enabling this can introduce a performance impact.
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Added from Abby's change

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx (hunk 15 above)
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Fixed

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Fixed by Abby's changes

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+defines a maximum amount of time that must exist between key clicks to be considered clicks rather than a key being
+held. This defaults to 0.1 seconds, and often should not have to be changed. If you do experience key click spamming,
+you can change this value to a higher number within the **Key Held Threshold Seconds** field.
+
+![Change the threshold for held key inputs](img/base_options_threshold.png)
+
+### Game object finder configurations
+
+The SDK traverses the game object tree of a game to collect state relevant for both recording playback and testing.
+In addition to ignore certain objects as described above, there are also a few other ways that this can be configured.
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: I believe this section has been removed / rearranged so this is no longer relevant

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx (hunk 16 above)
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Fixed!

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx (hunk 17 above)
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Added!

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx (hunk 18 above)
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Added!

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx (hunk 19 above)
[the remark this answers is not in the document]
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Added, let me know if that clarification is correct?

**Comment by vontell** [comment]
@RG-nAmKcAz back to you! I think I got them all, let me know if I am missing something, GitHub is not playing nicely with these comments on this PR...

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+
+## Advanced Options
+
+### Choosing an FPS for recording
+
+The `Recording Min FPS` field under `RGOverlayCanvasV2` > `ScreenRecorder` allows you to configure
+how often the SDK scrapes the game for state data.
+Its default value is 0, which instructs the SDK to only record state data when it detects a
+[**key frame**](/gameplay-sessions/gameplay-sessions-reference#key-frames). In most cases, this should be set to 0, as most games will only require state collection on key frames.
+Setting `Recording Min FPS` to a value greater than 0 will cause the SDK to record data at that frequency instead,
```
Comment (RG-nAmKcAz) [inline docs/gameplay-sessions/gameplay-sessions-usage.mdx:45]:
```suggestion
Setting `Recording Min FPS` to a value greater than 0 will cause the SDK to record data at that frequency,
```

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+regardless of whether a key frame was detected.
```
Comment (RG-nAmKcAz) [inline docs/gameplay-sessions/gameplay-sessions-usage.mdx:46]:
```suggestion
in addition to recording any key frame that was detected.
```

**Review by RG-nAmKcAz — APPROVED** [review]

**Comment by vontell** [comment]
@RG-nAmKcAz appreciate you committing those last two requests.
